apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommand.cs

```diff
@@ -219,6 +220,34 @@ public override UpdateRowSource UpdatedRowSource
             }
         }
 
+        // When query options are set, merge the new set of options into the existing options.
+        private void MergeQueryOptions(QueryOptions options)
+        {
+            // Nothing to merge.
+            if (options == null || string.IsNullOrEmpty(options.OptimizerVersion))
+            {
+                return;
+            }
+
+            if (_queryOptions == null)
+            {
+                _queryOptions = new QueryOptions();
+            }
+
+            _queryOptions.OptimizerVersion = options.OptimizerVersion;
+        }
+
+        /// <summary>
+        /// Query options to use when running SQL and streaming SQL commands.
+        /// When a new set of options is set, the new fields will be merged
+        /// into the fields in the existing options.
+        /// </summary>
+        public QueryOptions QueryOptions
```

Comment:
Right, this gets back to the earlier GetEffectiveOptions() discussion.

It feels a little odd that if I set the QueryOptions property and then fetch it again, I might not get the same result. The behavior of this code would be odd too:

```csharp
var command = new SqlCommand();
var options1 = new QueryOptions { OptimizerVersion = "foo" };
var options2 = new QueryOptions { OptimizerVersion = "bar" };

command.QueryOptions = options1;
var returnedOptions = command.QueryOptions;
Console.WriteLine(returnedOptions.OptimizerVersion); // foo

command.QueryOptions = options2;
Console.WriteLine(returnedOptions.OptimizerVersion); // bar
```

The mutation of an existing object feels odd.

It also feels odd that once I've set query options with an optimizer version, there's nothing I can do to clear it - setting QueryOptions to null doesn't clear the property itself, and setting it to a QueryOptions with an empty optimizer version has no effect.

I think it would be simpler to understand if each place where a QueryOptions can be specified had a simple automatically-implemented property with no logic, and we put the merging logic into the place where the request is constructed. We'd need to work out exactly what that merging logic *is*, admittedly - if the connection specifies a query optimizer version, can the command effectively remove that by specifying a non-null QueryOptions property, but with an empty query optimizer version? That would seem reasonably sensible to me.

Reply:
I hope the new implementation is consistent with your idea.

> We'd need to work out exactly what that merging logic is, admittedly - if the connection specifies a query optimizer version, can the command effectively remove that by specifying a non-null QueryOptions property, but with an empty query optimizer version?

I've asked the backend team on what the expected behavior is. But from the implementations for the other languages, we decided to go with the approach of only merge if a value is set. If a field/env var is null/empty then we assume that it was not set. To "unset" a value set at the connection level, you can explicitly set the default value. E.g. OptimizerVersion = "latest". This is what the backend assumes if no value is set.